zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/LearnerHandler.java

```diff
@@ -271,6 +280,8 @@ protected void setBufferedOutput(BufferedOutputStream bufferedOutput) {
             }
             throw new SaslException("Authentication failure: " + e.getMessage());
         }
+
+        this.messageTracker = new MessageTracker(BUFFERED_MESSAGE_SIZE);
```

Comment:
@hanm can you provide an option and disable this by default, I'm worried about the performance impact by tracking this, and probably we only enable this when we found there is potential consistency issue we want to debug.

Reply:
@lvfangmin On perf side, the recorded messages are bounded and the log will only be dumped when server disconnects, so the impact on performance would be negligible. That said, I think it's very reasonable to request a feature flag here. I'll add an option which by default turns all instrumentation and log dump into no-ops after I rebase this PR with latest master.

Reply:
@hanm I understand the logging part is low cost, but add/remove all the packets in the synchronized FIFO CircularBuffer is not trivial on high traffic environment, that's why I concerned about the performance.